### PR TITLE
convert.py: add python logging instead of print()

### PR DESCRIPTION
**Pull request by mofosyne**
Motivation of this PR is to switch from print() to python logging (battery included package for python), this is so that when I add the `--get-outfile` command to get the calculated conventional gguf filename it doesn't get polluted by erroneous print() messages.

### CODE REVIEW

**Comment by CISC** [comment]
Good suggestion, however you are using the logging module wrong. You must at the very least create and use a named instance with logging.getLogger(), see the [documentation](https://docs.python.org/3/library/logging.html#logger-objects).

**Comment by deiteris** [comment]
> Good suggestion, however you are using the logging module wrong. You must at the very least create and use a named instance with logging.getLogger(), see the [documentation](https://docs.python.org/3/library/logging.html#logger-objects).

The provided usage is not exactly wrong. You're not restricted from using library methods directly.

IMO, for a single python module (which is a standalone script here), it's fine to use default logger configuration and library methods if you don't need complex logging. But I'm not aware about the conventions here, so maybe other maintainers can say if it's ok or not.

**Comment by CISC** [comment]
> The provided usage is not exactly wrong. You're not restricted from using library methods directly.

While true it's a good idea not to, it will make things a lot easier in the long run. Just because you can doesn't mean you should. :)

> IMO, for a single python module (which is a standalone script here), it's fine to use default logger configuration and library methods if you don't need complex logging. But I'm not aware of conventions here, so maybe other maintainers can say if it's ok or not.

Ultimately it is up to whomever reviews this PR of course.

**Comment by mofosyne** [comment]
No probs. Made the adjustment. Hopefully it matches what you are thinking of

**Comment by CISC** [comment]
Almost. :)

You missed a couple of `logging` uses though, and you shouldn't use `__name__` as the logger name in a stand-alone script (it'll just be `__main__`), just calling it `convert` f.ex. will do fine.

**Comment by mofosyne** [comment]
> Almost. :)
> 
> You missed a couple of `logging` uses though, and you shouldn't use `__name__` as the logger name in a stand-alone script (it'll just be `__main__`), just calling it `convert` f.ex. will do fine.

You mean do_dump_model()? well in that case, its explicit a dump so I suppress all logging (unless --verbose is active) and only print whats on do_dump_model() happy to adjust behavior if that's unexpected.

**Comment by CISC** [comment]
> > You missed a couple of `logging` uses though, [...]
> You mean do_dump_model()? well in that case, its explicit a dump so I suppress all logging (unless --verbose is active) and only print whats on do_dump_model() happy to adjust behavior if that's unexpected.

No, I mean from [here](https://github.com/ggerganov/llama.cpp/pull/6511/files#diff-db4ed1f953cd7f07c37dfe2d725ac7f8ccae3fbb822cebd01662682f77e6ffe7R1523) and onwards.

**Comment by mofosyne** [comment]
> > > You missed a couple of `logging` uses though, [...]
> > > You mean do_dump_model()? well in that case, its explicit a dump so I suppress all logging (unless --verbose is active) and only print whats on do_dump_model() happy to adjust behavior if that's unexpected.
> 
> No, I mean from [here](https://github.com/ggerganov/llama.cpp/pull/6511/files#diff-db4ed1f953cd7f07c37dfe2d725ac7f8ccae3fbb822cebd01662682f77e6ffe7R1523) and onwards.

Ah dang. Yeah I see what you mean. Fixed thanks

**Comment by CISC** [comment]
> Ah dang. Yeah I see what you mean. Fixed thanks

Great, but [this](https://github.com/ggerganov/llama.cpp/pull/6511/commits/6e0ae011a050b86b9f6a71ba3d0fab086a06f2ea#diff-db4ed1f953cd7f07c37dfe2d725ac7f8ccae3fbb822cebd01662682f77e6ffe7R1482) was probably not intended though. :)

**Comment by mofosyne** [comment]
Fixed.

**Comment by rgryta** [comment]
Since logging is being introduced, we could also change `sys.stderr.write` call in the `default_outfile` method to `logger.error`.

I'm also wondering what'd be the process of ensuring that on later PR introduces new print statements to the code. Up for reviewers to detect, or add pipeline checks/tests - e.g. add `flake8-no-print` to the python-lint pipeline and replace `print` with `sys.stdout.write`.

**Comment by rgryta** [comment]
Nit: `\n` is now unnecessary in the logger.error 😅

**Comment by mofosyne** [comment]
> Since logging is being introduced, we could also change `sys.stderr.write` call in the `default_outfile` method to `logger.error`.
> 
> I'm also wondering what'd be the process of ensuring that later PRs don't introduce new print statements to the code. Up for reviewers to detect, or add pipeline checks/tests - e.g. add `flake8-no-print` to the python-lint pipeline and replace `print` with `sys.stdout.write`.

Alright then, I've figured out how to run flake8 directly (after failing to get pre-commit to work) and start using the flake8 no print allowed plugin. It then threw a whole bunch of errors on all the other pythons scripts in this repo. So decided to give correcting it all a stab.

Also updated the github actions and pre-commit config to use this feature so hopefully it now works.

Since this is changing quite a few scripts, I had to spend a bit more time on other .py scripts that had an older approach to using print() especially if ending with `end=` or not using fstrings for variables substitution. So you may want to pay attention to this when reviewing.

Unfortunately this has now make this PR bit more daunting to review compared to the past, but with the inclusion of the no-print check this should hopefully be a one time pain for everyone here. :)

File: .github/workflows/python-lint.yml [hunk 1]
```diff
@@ -22,3 +22,4 @@ jobs:
         with:
             ignore: "E203,E211,E221,E225,E231,E241,E251,E261,E266,E501,E701,E704,W503"
             exclude: "examples/*,examples/*/**,*/**/__init__.py"
```
Comment (cebtenzzre) [inline .github/workflows/python-lint.yml:24]: Does the GitHub runner not honor the flake8 configuration file (`.flake8`)?

File: convert-hf-to-gguf.py [hunk 2]
```diff
@@ -2461,13 +2464,19 @@ def parse_args() -> argparse.Namespace:
...
     return parser.parse_args()
 
 
 def main() -> None:
     args = parse_args()
 
+    if args.verbose:
+        logging.basicConfig(level=logging.DEBUG)
+    else:
+        logging.basicConfig(level=logging.INFO)
```
Comment (cebtenzzre) [inline convert-hf-to-gguf.py:2478]: This and most of the other basicConfig calls are a good candidate for a conditional expression.

```suggestion
    logging.basicConfig(level=logging.DEBUG if args.verbose else logging.INFO)
```

File: convert-llama-ggml-to-gguf.py [hunk 3]
```diff
@@ -125,7 +128,7 @@ def load(self, data, offset):
         self.start_offset = offset
         self.len_bytes = n_bytes
         offset += n_bytes
-        # print(n_dims, name_len, dtype, self.dims, self.name, pad)
+        # logger.info(n_dims, name_len, dtype, self.dims, self.name, pad)
```
Comment (cebtenzzre) [inline convert-llama-ggml-to-gguf.py:131]: remove this line

File: gguf-py/gguf/constants.py [hunk 4]
```diff
@@ -786,7 +789,7 @@ def get_type(val: Any) -> GGUFValueType:
             return GGUFValueType.INT32
         # TODO: need help with 64-bit types in Python
         else:
-            print("Unknown type:", type(val))
+            logger.error(f"Unknown type: {type(val)}")
             sys.exit()
```
Comment (cebtenzzre) [inline gguf-py/gguf/constants.py:793]: This module does not need a logger. This should really be:
```suggestion
            raise ValueError(f"Unknown type: {type(val)}")
```

File: gguf-py/scripts/gguf-convert-endian.py [hunk 5]
```diff
@@ -82,10 +85,11 @@ def convert_byteorder(reader: gguf.GGUFReader, args: argparse.Namespace) -> None
...
             if block_num % 100000 == 0:
-                print(f"[{(n_blocks - block_num) // 1000}K]", end="")
-                sys.stdout.flush()
-        print()
-    print("* Completion")
+                log_message += f"[{(n_blocks - block_num) // 1000}K]"
+
+        logger.info(log_message)
+
+    logger.info("* Completion")
```
Comment (cebtenzzre) [inline gguf-py/scripts/gguf-convert-endian.py:92]: This is clearly intended to be live progress output. If you are averse to using print for this (even with file=sys.stderr), use tqdm (which calls sys.stderr.write for you).

File: scripts/verify-checksum-models.py [hunk 6]
```diff
@@ -73,9 +76,9 @@ def sha256sum(file):
...
 # Print column headers for results table
-print("\n" + "filename".ljust(40) + "valid checksum".center(20) + "file missing".center(20))
-print("-" * 80)
+logger.info("filename".ljust(40) + "valid checksum".center(20) + "file missing".center(20))
+logger.info("-" * 80)
 
 # Output the results as a table
 for r in results:
-    print(f"{r['filename']:40} {r['valid checksum']:^20} {r['file missing']:^20}")
+    logger.info(f"{r['filename']:40} {r['valid checksum']:^20} {r['file missing']:^20}")
```
Comment (cebtenzzre) [inline scripts/verify-checksum-models.py:84]: This is the result of the program, it should be printed to stdout.

File: scripts/compare-llama-bench.py [hunk 7]
```diff
@@ -241,16 +238,12 @@ def get_commit_hexsha8(name):
...
         sys.exit(1)
 else:
-    print(
-        "ERROR: No compare target was provided and the current working directory "
-        "is not part of a git repository from which a compare target could be inferred."
-    )
-    print()
+    logger.error("No compare target was provided and the current working directory "
+                 "is not part of a git repository from which a compare target could be inferred.")
     parser.print_help()
```
Comment (cebtenzzre) [inline scripts/compare-llama-bench.py:247]: For readability, I think we should keep the blank line (`print()`) before calls to print_help. print_usage is short so it shouldn't matter.
```suggestion
    logger.error("No compare target was provided and the current working directory "
                 "is not part of a git repository from which a compare target could be inferred.\n")
    parser.print_help()
```

File: gguf-py/examples/reader.py [hunk 8]
```diff
@@ -18,28 +20,28 @@ def read_gguf_file(gguf_file_path):
...
-    print(tensor_info_format.format("Tensor Name", "Shape", "Size", "Quantization"))
-    print("-" * 80)
+    logger.info(tensor_info_format.format("Tensor Name", "Shape", "Size", "Quantization"))
+    logger.info("-" * 80)
     for tensor in reader.tensors:
         shape_str = "x".join(map(str, tensor.shape))
         size_str = str(tensor.n_elements)
         quantization_str = tensor.tensor_type.name
-        print(tensor_info_format.format(tensor.name, shape_str, size_str, quantization_str))
+        logger.info(tensor_info_format.format(tensor.name, shape_str, size_str, quantization_str))
```
Comment (cebtenzzre) [inline gguf-py/examples/reader.py:39]: All of this should be written to stdout (e.g. with `print()`), as it is the result of the program.

File: convert.py [hunk 9]
```diff
@@ -1466,8 +1469,18 @@ def main(args_in: list[str] | None = None) -> None:
...
     parser.add_argument("--skip-unknown", action="store_true",    help="skip unknown tensor names instead of failing")
+    parser.add_argument("--verbose",      action="store_true",    help="increase output verbosity")
 
     args = parser.parse_args(args_in)
+
+    if args.verbose:
+        logging.basicConfig(level=logging.DEBUG)
+    elif args.dump_single or args.dump:
+        # Avoid printing anything besides the dump output
+        logging.basicConfig(level=logging.CRITICAL)
```
Comment (cebtenzzre) [inline convert.py:1480]: Shouldn't we at least print warnings here? They would go to stderr and not hurt anything.

File: gguf-py/scripts/gguf-dump.py [hunk 10]
```diff
@@ -39,20 +42,21 @@ def dump_metadata(reader: GGUFReader, args: argparse.Namespace) -> None:
...
+                log_message += ' = {0}'.format(field.parts[-1][0])
+        logger.info(log_message)
     if args.no_tensors:
         return
-    print(f'\n* Dumping {len(reader.tensors)} tensor(s)')
+    logger.info(f'* Dumping {len(reader.tensors)} tensor(s)')
     for n, tensor in enumerate(reader.tensors, 1):
         prettydims = ', '.join('{0:5}'.format(d) for d in list(tensor.shape) + [1] * (4 - len(tensor.shape)))
-        print(f'  {n:5}: {tensor.n_elements:10} | {prettydims} | {tensor.tensor_type.name:7} | {tensor.name}')
+        logger.info(f'  {n:5}: {tensor.n_elements:10} | {prettydims} | {tensor.tensor_type.name:7} | {tensor.name}')
```
Comment (cebtenzzre) [inline gguf-py/scripts/gguf-dump.py:59]: This is the result of the program, so it should be written to stdout. `print()` seems like a perfectly acceptable way to do that.

File: gguf-py/scripts/gguf-convert-endian.py (hunk 5 above)
Comment (mofosyne) [reply gguf-py/scripts/gguf-convert-endian.py]: resolved, with specific focus on refactoring this section with emphases on why there is a special handling of the Q8_0 bock and to more easily add nested progress bars

**Comment by mofosyne** [comment]
rebased on top of latest master due to extra breaking changes on convert-hf-to-gguf.py (more print() was added in the main repo)

File: .github/workflows/python-lint.yml (hunk 1 above)
Comment (mofosyne) [reply .github/workflows/python-lint.yml:24]: I'm unsure, hence I'm not touching python-lint.yml ignore or exclude list.

I'm only updating .flake8 so that I can keep to the check on my side (as pre-commit isn't working for some reason) by directly calling `flake8`.

**Review by mofosyne — COMMENTED** [review]
Okay just finished making all the changes I noted from you (besides the flake8 enquiry which I'm taking as a comment for now).

I also took a bit of time to look over it again and it looks fine on my end so far.

**Comment by CISC** [comment]
Great work. :)

A final note however, you should use `__name__` as logger name for everything in gguf-py/gguf as they will be imported and not run directly.

**Comment by mofosyne** [comment]
> Great work. :)
> 
> A final note however, you should use `__name__` as logger name for everything in gguf-py/gguf as they will be imported and not run directly.

One of your previous statement appears to state the opposite.

> You missed a couple of logging uses though, and you shouldn't use __name__ as the logger name in a stand-alone script (it'll just be __main__), just calling it convert f.ex. will do fine.

So you are saying just specifically for `gguf-py/gguf` because it's used within other scripts?

File: .github/workflows/python-lint.yml (hunk 1 above)
Comment (cebtenzzre) [reply .github/workflows/python-lint.yml:24]: You could try removing it and see if the CI still passes on this PR.

Comment (mofosyne) [reply .github/workflows/python-lint.yml]: Well looks like it didn't break even after removing the ignores from this files etc... so looks like the .flake8 addition is taking up the slack now.

**Review by mofosyne — COMMENTED** [review]
Manually double checked and clicked viewed on all pages. Found one unused logger import and fixed it. Aside from that, looks ready.

**Comment by mofosyne** [comment]
Rebased to add one additional `[convert-hf-to-gguf.py: add additional logging](https://github.com/ggerganov/llama.cpp/pull/6511/commits/a6f54dee3ca65ae1dbeab3f8c26c1d75a9609715)` commit to fix addition of print() `convert-hf-to-gguf.py` in the main branch.

**Comment by teleprint-me** [comment]
I have a [pattern](https://github.com/teleprint-me/text-extraction/blob/main/text_extraction/logger.py#L13) we can use to centralize the logger instantiation with.

The implementation is really simple and flexible.

```python
def get_default_logger(name=None, level=logging.DEBUG):
    logger = logging.getLogger(name)
    if not logger.hasHandlers():
        handler = logging.StreamHandler(stream=sys.stdout)
        formatter = logging.Formatter(LOGGER_FORMAT)
        handler.setFormatter(formatter)
        logger.addHandler(handler)
        logger.setLevel(level)
    return logger
```

I left the docstring out of the snippet shown here, but this is basically it. You're free to use it to simplify the implementation and reduce code duplication. I've been able to use this in classes, functions, and scripts and regularly use it. It's under the MIT License, so it's compatible with `llama.cpp`.

This pattern can be extended over time and included in the `gguf-py` package as a `logger.py` module that can be improved over time.

**Comment by mofosyne** [comment]
Happy to add if @cebtenzzre thinks it's a good idea.

What's the advantage of a centralized logger instance? Would this be better as a separate PR (Which would be good if it's only a small change to each file).

The main objective of this PR is simply to convert from print() to logger in a preferably simple and extendable way, so at least for this one I would prefer to minimize adding dependencies.

**Comment by teleprint-me** [comment]
No worries. It was just a suggestion. No need to use it if undesired.

As for your questions.

This is builtin. No added dependency. You're already using it.

```python
import logging  # python builtin
```

Adding a default logger instance creates a predicable implementation, applies DRY, and reduces code duplication.

```
# Any name can be used here.
logger = get_default_logger(name="convert.py", level=logging.INFO)
```

Unified point of contact makes it so that a single change reflects the desired logging format from single location.

```python
LOGGER_FORMAT
```

The implementation details would be specific to `gguf-py`, hence the suggestion for a `logger.py` module that's customized to this projects specific needs.

You already made most of the changes, this would simply refine it. If you feel that this should be done in a separate PR, then so be it.

**Review by ggerganov — APPROVED** [review]
I assume this is OK to merge. Will give it another day for any additional feedback. @cebtenzzre Let us know if you need some more time to review

**Comment by mofosyne** [comment]
Hopefully all is good? Noticed more changes to base branch so did a local rebase and flake8 to check for any issues. So far no problems so no force pushing needed.

<details>
```
$ git fetch upstream master
remote: Enumerating objects: 173, done.
remote: Counting objects: 100% (148/148), done.
remote: Compressing objects: 100% (110/110), done.
remote: Total 173 (delta 87), reused 71 (delta 38), pack-reused 25
Receiving objects: 100% (173/173), 815.29 KiB | 931.00 KiB/s, done.
Resolving deltas: 100% (88/88), completed with 7 local objects.
From github.com:ggerganov/llama.cpp
 * branch            master     -> FETCH_HEAD
 * [new branch]      master     -> upstream/master
$ git rebase upstream master
fatal: invalid upstream 'upstream'
$ git rebase upstream/master
Successfully rebased and updated refs/heads/convert-py-logging.
$ flake8
```
</details>

File: tests/test-tokenizer-0-falcon.py [hunk 11]
```diff
@@ -1,14 +1,19 @@
 # tests with BPE tokenizer
 
+import logging
 import argparse
 
 from transformers import AutoTokenizer
 
+logger = logging.getLogger("convert")
```
Comment (CISC) [inline tests/test-tokenizer-0-falcon.py:8]: Wrong name.

**Comment by mofosyne** [comment]
Github noticed some new conflicts so manually rebased and resolved the conflicts.

Also then noticed the addition of this new file `convert-hf-to-gguf-update.py` which is triggering warnings in flake8 when removed from ignore list. Decided to resolve all flake8 warnings manually so there be some extra changes to `convert-hf-to-gguf-update.py`.

Thanks for the catch, hopefully this is closer!

File: convert-hf-to-gguf-update.py [hunk 12]
```diff
@@ -115,76 +123,75 @@ def download_file_with_auth(url, token, save_path):
         continue
 
     # create the tokenizer
-    from transformers import AutoTokenizer
```
Comment (mofosyne) [inline convert-hf-to-gguf-update.py:118]: Moved to the top to stop a warning

**Comment by mofosyne** [comment]
Resolved conflict in two files.

File: convert-hf-to-gguf-update.py [hunk 13]
```diff
@@ -115,76 +123,75 @@ def download_file_with_auth(url, token, save_path):
...
-src_func +=  "        print(f\"chkhsh: {chkhsh}\")\n"
-src_func +=  "\n"
-src_func +=  "        return res\n"
-
-print(src_func)
-
-print("\n")
-print("!!! Copy-paste the function above into convert-hf-to-gguf.py !!!")
-print("\n")
+src_func = "" # noqa: E222
```
Comment (CISC) [inline convert-hf-to-gguf-update.py:149]: It would be better to use f"""-string here (then you wouldn't have to `noqa` every line either), but it might be overreaching for this PR? @ggerganov ?

File: convert-hf-to-gguf.py [hunk 14]
```diff
@@ -1195,10 +1186,9 @@ def write_tensors(self):
             data = data_torch.squeeze().numpy()
 
             # map tensor names
-            new_name = tensor_map.get_name(name, try_suffixes=(".weight",))
+            new_name = tensor_map.get_name(name, try_suffixes=(".weight", ".bias"))
```
Comment (CISC) [inline convert-hf-to-gguf.py:1189]: Undo this change.

**Comment by CISC** [comment]
@mofosyne Great work here, unfortunately I think you were a bit unlucky and fell right between some urgent events, leaving everyone very busy, which has postponed the merge of this PR.

Also I do not have any reviewer powers, just being a busy-body. :) You'll have to wait for @cebtenzzre and/or @ggerganov to get some spare time.

**Review by CISC — COMMENTED** [review]
Hate to do this to you, but you got the wrong one, the correct one is at L1189, please revert 6d42f3d7736ab74b44057e53995b0e1e1627efb6 :)

**Comment by mofosyne** [comment]
> @mofosyne Great work here, unfortunately I think you were a bit unlucky and fell right between some urgent events, leaving everyone very busy, which has postponed the merge of this PR.
> 
> Also I do not have any reviewer powers, just being a busy-body. :) You'll have to wait for @cebtenzzre and/or @ggerganov to get some spare time.

Ah that's not a problem. Ultimately, this is mostly just a quality of life fix to make developer life smoother. But I do hope to eventually get this in, as better debugging will ultimately make everyone's life easier.

File: tests/test-tokenizer-0-spm.py [hunk 15]
```diff
@@ -65,41 +72,46 @@
...
-print("'" + tokenizer.decode([15043, 15043]) + "'") # 'Hello Hello'
-print("'" + tokenizer.decode([29871, 15043]) + "'")               # ' Hello'
-print("'" + tokenizer.decode([29871, 15043, 29871, 15043]) + "'") # ' Hello  Hello'
-
-print("\n\ntests for C++:\n")
+    message_log = (f"text: {text}\n"
+                   "with bos:\n"
+                   f"{tokenizer.encode(text, add_bos=True)}\n"
+                   f"{tokenizer.decode(tokenizer.encode(text, add_bos=True))}\n"
+                   "without bos:\n"
```
Comment (compilade) [inline tests/test-tokenizer-0-spm.py:79]: This removes the additional newlines before `"with bos"` and `"without bos"`. Just making sure you intended to remove this.

File: requirements.txt [hunk 16]
```diff
@@ -5,7 +5,6 @@
 #
 
 -r ./requirements/requirements-convert.txt
-
```
Comment (compilade) [inline requirements.txt:8]: I think it was previously separated because it's included in most of the other requirements files below, to make it stand out.

File: convert-hf-to-gguf.py [hunk 17]
```diff
@@ -2721,7 +2691,7 @@ def set_vocab(self):
         else:
             # Use the GPT-NeoX tokenizer when no tokenizer files are present
             tokenizer_path = Path(sys.path[0]) / "models" / "ggml-vocab-gpt-neox.gguf"
-            print(f"Using tokenizer from '{os.path.relpath(tokenizer_path, os.getcwd())}'")
+            logger.debug(f"Using tokenizer from '{os.path.relpath(tokenizer_path, os.getcwd())}'")
```
Comment (compilade) [inline convert-hf-to-gguf.py:2694]: This doesn't happen often and I think the user should know

```suggestion
            logger.warning(f"Using tokenizer from '{os.path.relpath(tokenizer_path, os.getcwd())}'")
```

File: convert-hf-to-gguf.py [hunk 18]
```diff
@@ -2390,7 +2363,7 @@ def set_vocab(self):
             # TODO: this is a hack, should be fixed
             #       https://github.com/ggerganov/llama.cpp/pull/6745#issuecomment-2067687048
             special_vocab.special_token_ids["eos"] = self._try_get_sft_eos(tokenizer)
-            print(f"Replace eos:{old_eos} with a special token:{special_vocab.special_token_ids['eos']} \
+            logger.debug(f"Replace eos:{old_eos} with a special token:{special_vocab.special_token_ids['eos']} \
```
Comment (compilade) [inline convert-hf-to-gguf.py:2366]: I think this should also be more prominent, because the comment above says "this is a hack, should be fixed"

```suggestion
            logger.warning(f"Replace eos:{old_eos} with a special token:{special_vocab.special_token_ids['eos']} \
```

**Review by compilade — COMMENTED** [review]
This is overall quite good. Feel free to ignore/"resolve" any of the comments I made if they are too nit-picky.

**Comment by mofosyne** [comment]
Okay implemented most of your suggestions @cebtenzzre but excluded some style observations as it doesn't appear to be an issue.

**Review by compilade — APPROVED** [review]
Looks good to me. Merging this will cause conflicts in at least #7041, #7036, #7033, #7031, #7027, #7018, #6942, #6866 and soft conflicts with at least #7020, #6986, #6919, #6826. (soft conflicts in this case mean new linter errors, but no merge conflict)

Those involved in these PRs will probably want to take at least a cursory look at this before then (hopefully soon).

**Comment by teleprint-me** [comment]
Maybe this one should be prioritized for now?

**Comment by mofosyne** [comment]
Thanks for merging, hopefully all is well for other PRs!

**Comment by ggerganov** [comment]
On my Macbook I don't see the print's anymore. For example:

```bash
python3 convert-hf-to-gguf-update.py
```

should print usage, but after this PR it does not print anything. Any ideas?

**Comment by mofosyne** [comment]
I see what's going on. By default in https://docs.python.org/3/howto/logging.html logging level is only warning and up.

> The default level is WARNING, which means that only events of this level and above will be tracked, unless the logging package is configured to do otherwise.

To get it printing again you need to add `logging.basicConfig(level=logging.DEBUG)` just under `logger = logging.getLogger("convert-hf-to-gguf-update")`

This was missed because it was not using the typical `parser.add_argument` that I was searching for.

**Comment by CISC** [comment]
> Hmm I wonder why not all f-strings have been converted to %s parameters like all logging libraries suggest for a _bit_ lazier evaluation 😅

If it would impact performance in any meaningful way, sure, otherwise no. IMO f-strings are much better and less error prone.

**Comment by slaren** [comment]
This has broken `compare-llama-bench.py`.

**Comment by ggerganov** [comment]
> This has broken `compare-llama-bench.py`.

@mofosyne PTAL - usage instructions are here: https://github.com/ggerganov/llama.cpp/blob/c780e75305dba1f67691a8dc0e8bc8425838a452/scripts/compare-llama-bench.py#L45-L57
